Design/spectrum_number_and_workspace_index_abstraction.md

```diff
@@ -0,0 +1,269 @@
+- Spectrum numbers can be defined differently in each workspace.
+  As a consequence the object dealing with translation must be linked to a workspace.
+  For example, it could be a member of the workspace.
+- We should make sure that we do not negatively influence performance when a trivial set of indices is specified by the user.
+  Trivial can mean either the full range or a range in combination with a trivial mapping from spectrum number to workspace index (such as a simple offset of 1).
+  I think our existing performance tests are in general not sensitive enough for this.
+  However, it will probably be fine if we support some special cases (which are actually the most common ones):
+
+  - We want to avoid going through a `map` or `vector` for translation in the simple cases of "full-range" (all spectra), or a simple "min-max" case.
+    This is simple to do, an example can be found as part of [this pull request](https://github.com/mantidproject/mantid/pull/15465)
+    I believe this will cover 95% of cases and ensure that there will be no performance hit.
+  - In the other cases, when an actual list of spectra is given, I suppose that that list is commonly small (since it is often manual user input), so `map` access is negligible.
+  - The remaining performance critical case is loading event data. Each event has an associated detector ID, and this needs to be translated to a workspace index.
+    This is currently done by creating a `vector` (instead of a `map`) for look-up.
+    We certainly have to support a similar mechanism.
+    This is a case where we will need to write some performance tests, to prevent regression.
+    These tests should be added well in advance so we can gather enough statistics prior to any changes and monitor the introduction of the changes.
+
+To deal with MPI in a way that is hidden from both the user and most developers, `Translator` transforms a user input into, e.g., a `SpectrumIndexSet`.
+This is best illustrated by the example from above:
+
+```python
+# User input: SpectrumNumbers='2,3,7'
+indices = ws.makeSpectrumIndexSet(getProperty('SpectrumNumbers'))
+```
+
+Consider 3 cases:
+
+```python
+# Without MPI (or single MPI rank)
+Spectrum number        1 2 3 7 8
+Workspace index        0 1 2 3 4
+Global workspace index 0 1 2 3 4
+# indices contains 1,2,3
+
+# 2 MPI ranks (interleaved distribution of spectra)
+Spectrum number        1 2 3 7 8
+MPI rank               0 1 0 1 0
+Workspace index        0 0 1 1 2
+Global workspace index 0 1 2 3 4
+# rank 0: indices contains 1
+# rank 1: indices contains 0,1
+
+# 3 MPI ranks (block-wise distribution of spectra)
+Spectrum number        1 2 3 7 8
+MPI rank               0 0 1 1 2
+Workspace index        0 1 0 1 0
+Global workspace index 0 1 2 3 4
+# rank 0: indices contains 1
+# rank 1: indices contains 0,1
+# rank 2: indices is empty
+```
+
+Algorithms then use:
+
+```python
+for index in indices:
+  ws.mutableY(index) += offset
```

Comment:
`mutableY(index)` returns a `HistogramY` at that workspace index. `HistogramY` is iterable, so you can do:

```cpp
for (const auto & y : ws.mutableY(index))
  sumY += y;
```
We have also considered avoiding the explicit loop over `indices` by providing something like `MatrixWorkspace::apply(const SpectrumIndexSet &indices, T operator)` (or maybe as part of `Algorithm`, so we can do cancelling and progress reporting).